contracts/EarningOracle.sol

```diff
@@ -0,0 +1,36 @@
+// SPDX-License-Identifier: agpl-3.0
+pragma solidity ^0.6.12;
+pragma experimental ABIEncoderV2;
+
+import "./interfaces/IAdExCore.sol";
+import "./interfaces/IEarningOracle.sol";
+import "./libs/ChannelLibrary.sol";
+
+
+contract EarningOracle is IEarningOracle {
+    IAdExCore public core;
+
+    mapping (address => uint) internal totalEarnings;
+    mapping (address => mapping(bytes32 => bool)) public tallied;
+
+    constructor (IAdExCore _core) public {
+        core = _core;
+    }
+
+    function bulkUpdate(bytes32[] calldata channelIds, address[] calldata earners) external {
+        for(uint i = 0; i < channelIds.length; i++) {
+            bytes32 channelId = channelIds[i];
+            address earner = earners[i];
+            
+            require(tallied[earner][channelId] == false, 'ALREADY_TALLIED');
```

Comment:
Consider using `if` instead `require`. With `require` you have to keep strict state of the updated channels. For example - there are 10 channels on the first bulk update, after some time there are 5 new channels and it is time to make new `bulkUpdate` but you don't keep record of the previous updated channels and make update with all 15 channels - and the update will be "reverted".

Reply:
The record of previously updated channels will be kept off-chain.

Reply:
Ok